csrc/multidevice/device_mesh.h

```diff
@@ -67,20 +85,42 @@ class DeviceMesh final {
     return -1;
   }
 
-  // Returns the device at a particular index in the mesh
+  // Returns the indices of a multi-dimensional mesh, or an empty vector 
+  // if device is not present
+  std::vector<int64_t> getLocalIndices(const DeviceIdxType device) const;
+
+  // Returns the device at a particular global index in the mesh
   DeviceIdxType at(int64_t index) const {
     return vector_.at(index);
   }
 
   bool operator==(const DeviceMesh& other) const {
-    return vector_ == other.vector();
+    return vector_ == other.vector() && shape_ == other.shape();
   }
 
+  // Returns the max device index in the DeviceMesh.
+  DeviceIdxType maxDeviceIdx() const;
```

Comment:
If a method can be done without accesses to private fields, I tend to make it a helper function on the side. This way, we minimize the exposure of private fields so it's easier to change them in the future. 

```suggestion
DeviceIdxType maxDeviceIdx(const DeviceMesh&);
```